parasol/query.py

```diff
@@ -0,0 +1,241 @@
+"""
+Object-oriented approach to Solr searching and filtering modeled
+on :class:`django.models.queryset.QuerySet`.  Supports iteration,
+slicing, counting, and boolean check to see if a search has results.
+
+Filter, search and sort methods return a new queryset, and can be
+chained. For example::
+
+    SolrQuerySet().filter(item_type='person') \
+                  .search(name='hem*') \
+                  .order_by('sort_name') \
+
+
+"""
+
+from typing import Any, Optional
+
+try:
+    from parasol.solr.django import SolrClient
+except ImportError:
+    # FIXME: SolrQuerySet doesn't currently work with non-django
+    # solr client because it would need a way to get connection settings
+    from parasol.solr import SolrClient
+
+
+class SolrQuerySet:
+    """A Solr queryset object that allows for object oriented
+    searching and filtering of Solr results. Allows search results
+    to be paginated by django paginator."""
+
+    _result_cache = None
+    start = 0
+    stop = None
+    sort_options = []
+    search_qs = []
+    filter_qs = []
+
+    #: by default, combine search queries with AND
+    default_search_operator = 'AND'
```

Comment:
Not for now at all, but this might be a good candidate for more complicated logic with something analogous to Django's `Q` syntax. (Also worth looking at `elasticsearch-dsl` on this, as they have a similar configuration.)

Reply:
Agreed, we might need that to support the queries we're doing now for the PPA archive search.